core/runtime/impl/block_builder_impl.cpp

```diff
@@ -0,0 +1,139 @@
+      : memory_{extension->memory()},
+        codec_{std::move(codec)},
+        executor_{std::move(extension)},
+        state_code_{std::move(state_code)} {}
+
+  outcome::result<bool> BlockBuilderImpl::apply_extrinsic(Extrinsic extrinsic) {
+    OUTCOME_TRY(encoded_ext, codec_->encodeExtrinsic(extrinsic));
+
+    runtime::SizeType ext_size = encoded_ext.size();
+    runtime::WasmPointer ptr = memory_->allocate(ext_size);
+    memory_->storeBuffer(ptr, encoded_ext);
+
+    wasm::LiteralList ll{Literal(ptr), Literal(ext_size)};
+
+    OUTCOME_TRY(
+        _, executor_.call(state_code_, "BlockBuilder_apply_extrinsic", ll));
+
+    return true;
+    /// @todo figure out what wasm function returns
+  }
+
+  outcome::result<primitives::BlockHeader> BlockBuilderImpl::finalize_block() {
+    OUTCOME_TRY(res,
+                executor_.call(state_code_, "BlockBuilder_finalize_block", {}));
+
+    uint32_t res_addr = static_cast<uint64_t>(res.geti64()) >> 32ul;
+
+    WasmMemoryStream s(memory_);
+    OUTCOME_TRY(s.advance(res_addr));
+
+    OUTCOME_TRY(header, codec_->decodeBlockHeader(s));
+
+    return std::move(header);  // warning from clang-tidy without move
+  }
+
+  outcome::result<std::vector<primitives::Extrinsic>>
+  BlockBuilderImpl::inherent_extrinsics(primitives::InherentData data) {
+    OUTCOME_TRY(enc_data, codec_->encodeInherentData(data));
+
+    runtime::SizeType data_size = enc_data.size();
+    runtime::WasmPointer ptr = memory_->allocate(data_size);
+    memory_->storeBuffer(ptr, enc_data);
+
+    wasm::LiteralList ll{Literal(ptr), Literal(data_size)};
+
+    OUTCOME_TRY(
+        res,
+        executor_.call(state_code_, "BlockBuilder_inherent_extrinsics", ll));
+
+    uint32_t res_addr = static_cast<uint64_t>(res.geti64()) >> 32ul;
+
+    WasmMemoryStream s(memory_);
+    OUTCOME_TRY(s.advance(res_addr));
+
+    using std::placeholders::_1;
+    auto decode_f = [this](common::ByteStream &stream) {
+      return codec_->decodeExtrinsic(stream);
+    };
+    OUTCOME_TRY(decoded_res, decodeCollection<Extrinsic>(s, decode_f));
+
+    return std::move(decoded_res);
+  }
+
+  outcome::result<CheckInherentsResult> BlockBuilderImpl::check_inherents(
+      primitives::Block block, primitives::InherentData data) {
+    OUTCOME_TRY(enc_block, codec_->encodeBlock(block));
+    OUTCOME_TRY(enc_data, codec_->encodeInherentData(data));
+
+    runtime::SizeType block_size = enc_block.size();
+    runtime::WasmPointer block_ptr = memory_->allocate(block_size);
+    memory_->storeBuffer(block_ptr, enc_block);
+
+    runtime::SizeType data_size = enc_data.size();
+    runtime::WasmPointer data_ptr = memory_->allocate(data_size);
+    memory_->storeBuffer(data_ptr, enc_data);
+
+    wasm::LiteralList ll{Literal(block_ptr), Literal(data_ptr)};
+
+    OUTCOME_TRY(
+        res, executor_.call(state_code_, "BlockBuilder_check_inherents", ll));
+
+    uint32_t res_addr = static_cast<uint64_t>(res.geti64()) >> 32ul;
```

Comment:
I think, we should create a function or macros for this transformation, as it is repeated a lot of times

Reply:
you mean " >> 32ul" ?

Reply:
With static cast, yes, it's used a lot.